mixed dtype linear

mikekgfb opened the pull request:
Summary: mixed dtype linear

Test Plan:
Looking to support the same mixed dtype operation as GPU

```
buck test mode/opt mode/inplace :test_matmul_new_cpu
```

Differential Revision: D53209472




cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

mikekgfb commented:
@alexsamardzic please have a look what needs to change to make this work on CPU.  We suspect this relates primarily to the swizzling of the layout to accommodate GPU layout?  
Would love if you could take this over.  (I commented out int4 for now, but we'll need to support the int4 as well)

cc: @cpuhrsch @malfet

cpuhrsch commented on test/test_matmul_new_cpu.py:
```diff
@@ -0,0 +1,150 @@
+# Owner(s): ["module: linear algebra"]
```
It'll be easier to land this if this is part of test_torch.py.

alexsamardzic commented:
Attached is a patch that will make the test to pass; basically, all GPU related quirks are removed, file `aten/src/ATen/native/cpu/MixedDtypesLinear.cpp` is moved one directory above (I never implemented CPU version of an ATen operator before, and it seems there are some specific things to do in order to be able to put a file into this `cpu` subdirectory, otherwise build will fail), and I introduced accumulator in higher precision to have results to match with non-quantized case.

The main problem here is however that this is going to be rather slow.  Namely, an accelerated GEMM library is typically to be used both for CPU and GPU.  For GPU case, corresponding libraries do not yet provide support for mixed data-types GEMM off-the-shelf, so our approach was to use CUTLASS open-source library, and extend it with the functionality needed (I'm actually still working on this).  For CPU, I'm not sure exactly what PyTorch uses, I can see references to MKL and OpenBLAS in the code, I don't know if it uses some additional libraries.  As far as I can see, these two libraries do not support mixed data-types GEMM yet.  There is [BLIS](https://github.com/flame/blis) library available that seems to support it, but I'm not sure is it worth making it a dependency for this reason only, nor that its performance is going to be satisfactory vs. just up-casting and using existing multiplication operators.

[patch.txt](https://github.com/pytorch/pytorch/files/14102062/patch.txt)

cpuhrsch commented:
@alexsamardzic - Thanks for the review! You should be able to add this patch by pushing to this branch or just opening a new PR with it in included. One advantage of this plain CPU code is also that it can be compiled for different architectures. The accelerator libraries you mention are optimized for x86.

Commit:
mixed dtype linear (#118606)

Summary:
mixed dtype linear


Test Plan:
Looking to support the same mixed dtype operation as GPU

```
buck test mode/opt mode/inplace :test_matmul_new_cpu
```





cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

Differential Revision: D53209472

Pulled By: cpuhrsch

alexsamardzic commented:
I've made some additional changes (including enabling bias), and pushed my commit.  There are number of fixes/extensions to be made, that I've marked with `FIXME`s in the code, but the most important here is:

1. To decide upon operands data-types and layouts combinations to be supported.  (Along with this, it is to be decided about actual quantization schemes to be supported - this is something that we have to do for the GPU version too).
2. To still change the matrix multiplication algorithm.  CPU matrix multiplication is not exactly my cup of tea, but keeping it this way really doesn't make sense w.r.t. performance.  I'm not sure which CPU architectures PyTorch is supposed to support, but both OpenBLAS and BLIS support number of them, and if there is mismatch, this should be at least implemented in OpenMP, for multi-core support, and with some sort of blocking put in place, in order to properly utilize caches.

lezcano commented:
I agree with @alexsamardzic's here. If we want this to at least have a reasonable performance, we should implement some optimisations like
